### ISSUE TITLE
11-image-classifier-dogs-vs-cats.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/elius123ef#9]
## LO QUE HAS HECHO BIEN

- **Importaciones completas**: tensorflow, keras, ImageDataGenerator, Conv2D, MaxPool2D, Dense, callbacks
- **Seed fijada** para reproducibilidad
- **Visualización de imágenes**: primeras 9 de dogs y cats correctamente mostradas
- **Normalización de imágenes** al visualizar (`/= 255.0`)
- **ImageDataGenerator** con `flow_from_directory` - estructura correcta
- **Arquitectura CNN bien diseñada**: 
  - `BatchNormalization` (mejora convergencia)
  - `GlobalAveragePooling2D` (reduce overfitting vs Flatten)
  - `Dropout(0.5)` (regularización)
- **Compilación correcta**: `categorical_crossentropy`, Adam con lr=1e-4
- **Callbacks completos**: EarlyStopping, ReduceLROnPlateau, ModelCheckpoint 
- **Gráficas de accuracy y loss** para train y validation
- **Predicción con imagen de prueba**

### 1. **Falta normalización en ImageDataGenerator**
```python
# No normalizas las imágenes en el generator
datagentrain = ImageDataGenerator()
datagentest  = ImageDataGenerator()

# Deberías normalizar (rescale) y añadir data augmentation:
datagentrain = ImageDataGenerator(
    rescale=1./255,
    rotation_range=20,
    width_shift_range=0.2,
    height_shift_range=0.2,
    horizontal_flip=True,
    zoom_range=0.2
)

datagentest = ImageDataGenerator(rescale=1./255)  # Solo rescale para test
```
**¿Por qué?** Sin `rescale=1./255`, el modelo recibe valores 0-255 en lugar de 0-1. Las redes neuronales convergen mejor con valores pequeños y las funciones de activación están diseñadas para ese rango.

### 2. **Error de ruta en la predicción (FALTA UNA BARRA)**

<img width="967" height="551" alt="Image" src="https://github.com/user-attachments/assets/b5bcc271-ef43-42b6-b005-a70914ec066f" />

Falta el backslash después de los dos puntos

```py
# Correcto:
img = image.load_img(r"..\data\Test\Cats\cat.10000.jpg", ...)
saved_model = load_model(r"..\models\best_model.h5")
```

`..data` busca una carpeta llamada "..data". `..\data` sube un nivel y entra a `data`.

### 3. **Falta normalizar la imagen al predecir**
```python
# La imagen no se normaliza antes de predecir
img = np.asarray(img)
img = np.expand_dims(img, axis=0)
output = saved_model.predict(img)  # ← Recibe valores 0-255!

# Debes normalizar igual que en entrenamiento:
img = np.asarray(img) / 255.0  # ← Normalizar!
img = np.expand_dims(img, axis=0)
output = saved_model.predict(img)
```
Si entrenas con valores 0-1 pero predices con 0-255, el modelo no reconocerá la imagen.

### 4. **ModelCheckpoint guarda en directorio incorrecto**
```python
# Guarda en el directorio actual
ModelCheckpoint("best_model.h5", ...)

# Debería guardar en ../models/:
ModelCheckpoint("../models/best_model.h5", ...)
```

### 5. **Faltan métricas de evaluación finales**
```python
# Después de entrenar, añade evaluación formal:
loss, accuracy = model.evaluate(test_data)
print(f"Test Loss: {loss:.4f}")
print(f"Test Accuracy: {accuracy:.4f}")
```

---

Buen trabajo @elius123ef . La arquitectura está bien pensada con BatchNormalization, Dropout y GlobalAveragePooling. Los puntos críticos a corregir:

1. **Añade `rescale=1./255`** en ImageDataGenerator - sin esto el modelo recibe valores 0-255 en vez de 0-1
2. **Corrige las rutas** en la última celda: `..\data` en lugar de `..data`
3. **Normaliza la imagen** al predecir: `img / 255.0`

**Comment by elius123ef** [comment]
Corregido
